src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs

```diff
@@ -182,24 +174,36 @@ private async Task<Solution> UpdateProjectsAsync(Solution solution, HashSet<Chec
                 }
 
                 return solution;
+
+                async Task PopulateOldProjectMapAsync()
+                {
+                    foreach (var (projectId, projectState) in solution.State.ProjectStates)
+                    {
+                        var projectChecksums = await projectState.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
+                        if (olds.Object.Contains(projectChecksums.Checksum))
+                            oldMap.Add(projectId, projectChecksums);
+                    }
+                }
+
+                async Task PopulateNewProjectMapAsync(SolutionCreator @this)
+                {
+                    var projectStateChecksums = await @this._assetProvider.GetAssetsAsync<ProjectStateChecksums>(news.Object, cancellationToken).ConfigureAwait(false);
+
+                    foreach (var (_, projectStateChecksum) in projectStateChecksums)
+                        newMap.Add(projectStateChecksum.ProjectId, projectStateChecksum);
```

Comment:
this code is much more efficient than what is is replacing.  because the projectStateChecksums exposes a .ProjectId now, we avoid a pointless roundtrip between the server and the host.